File: CompareVersionNumbers/compare_version_numbers.py
class Solution:
    def remove_tail_zero(self, v_list):
        v_list_len = len(v_list)
        for i in xrange(v_list_len-1, -1, -1):
            if v_list[i] == 0:
                v_list.pop()
            else:
                break
        if len(v_list) == 0:
            return [0]
        return v_list

    # @param version1, a string
    # @param version2, a string
    # @return an integer
    def compareVersion(self, version1, version2):
        v1_list = version1.split('.')
        v2_list = version2.split('.')
        v1_list = map(int, v1_list)
        v2_list = map(int, v2_list)
        v1_list = self.remove_tail_zero(v1_list)
        v2_list = self.remove_tail_zero(v2_list)

        for i, v in enumerate(v1_list):
            try:
                if v > v2_list[i]:
                    return 1
                elif v < v2_list[i]:
                    return -1
            except IndexError:
                return 1
        if len(v2_list) > len(v1_list):

            return -1
        return 0

s = Solution()
print(s.compareVersion('1.2', '1'))  # 1
print(s.compareVersion('1', '1.2'))  # -1
print(s.compareVersion('1.0', '1'))  # 0
print(s.compareVersion('1.2.3', '2'))  # -1
print(s.compareVersion('2', '1.2.3'))  # 1
print(s.compareVersion('1.2.3', '1.2.4'))  # -1
print(s.compareVersion('1.2.3', '1.3.4'))  # -1
print(s.compareVersion('0.0.0', '0'))  # 0
print(s.compareVersion('0', '0'))  # 0
print(s.compareVersion('0', '0.0.0.1'))  # -1
print(s.compareVersion('0.0.0.0.2', '0.0.0.1'))  # -1
print(s.compareVersion('0.2.0.0.2', '0.0.0.1'))  # 1

File: CompareVersionNumbers/compare_version_numbers.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    static int POW10[];

    int stringToInt(const string& str) {
        int str_len = str.length();
        int i, ret = 0;
        for (i=0; i<str_len; i++) {
            ret += (str[i]-'0') * (Solution::POW10[str_len-i-1]);
        }
        return ret;
    }

    vector<string> split(const string& str, const string& sep) {
        int index, sep_len = sep.length(), left = 0;
        vector<string> ret = vector<string>();
        while ((index = str.find(sep, left)) != -1) {
            ret.push_back(str.substr(left, index-left));
            left = index + sep_len;
        }
        ret.push_back(str.substr(left));
        return ret;
    }

    vector<string> split(const string& str, char sep) {
        int index, left = 0;
        vector<string> ret = vector<string>();
        while ((index = str.find(sep, left)) != -1) {
            ret.push_back(str.substr(left, index-left));
            left = index + 1;
        }
        ret.push_back(str.substr(left));
        return ret;
    }

    void remove_tail_zero()

    int compareVersion(string version1, string version2) {
        vector<string> v1_list = this->split(version1, '.');
        vector<string> v2_list = this->split(version2, '.');

        return 0;
    }
};

int Solution::POW10[] = {1, 10, 100, 1000, 10000, 100000, 1000000, 10000000};

void printVector(const vector<string>& v)
{
    for (vector<string>::const_iterator it=v.begin(); it!=v.end(); it++) {
        cout << *it << endl;
    }
}

int main(void)
{
    Solution s = Solution();
    int i = s.stringToInt(string("123"));
    printVector(s.split(string("1,2,3,4"), ','));
    cout << i;
    return 0;
}
